### PR TITLE
matter end-device implemented in Python (v2)

### PR DESCRIPTION
**Pull request by markus-becker-tridonic-com**
#### Problem

Currently there is a Python matter controller in the source tree. There is no Python matter end-device.

v1 of this PR is in https://github.com/project-chip/connectedhomeip/pull/8589, but cannot be reopened.

#### Change overview

This PR facilitates to create a Python based matter end-device.

* Adds python/chip/server/ServerInit.cpp and Python wrapper
* Adds Python lighting device sample with README in examples/lighting-app/python
  * This sample requires a DALI-USB-interface
* Checked onnetwork / ip commissioning and BT/WiFi commissioning
* In the initiial version of this PR a modified controller-clusters.zap was used. Now the correct lighting-app.zap is used.
* build-chip-wheel.py was auto-formatted

#### Testing

* Manual test of the new Python matter end-device against chip-device-ctrl running on the same host computer as described in the README of the sample added with this PR.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Review by bzbarsky-apple — DISMISSED** [review]
>     3rd endpoint in lighting-app is removed.

Why is this OK?

**Review by electrocucaracha — COMMENTED** [review]
I noticed several code lines commented, are these leftovers or they have some future purpose?

File: src/app/chip_data_model.gni [hunk 1]
```diff
@@ -149,14 +150,19 @@ template("chip_data_model") {
...
       sources += [
-        "${invoker.zap_pregenerated_dir}/CHIPClusters.cpp",
-        "${invoker.zap_pregenerated_dir}/CHIPClusters.h",
-        "${invoker.zap_pregenerated_dir}/CHIPClustersInvoke.cpp",
-        "${invoker.zap_pregenerated_dir}/CHIPClustersWrite.cpp",
         "${invoker.zap_pregenerated_dir}/attribute-size.cpp",
         "${invoker.zap_pregenerated_dir}/callback-stub.cpp",
       ]
 
+      if (chip_python_controller) {
```
Comment (bzbarsky-apple) [inline src/app/chip_data_model.gni:157]: This seems odd.  Why do other things that use controller clusters, say, not need this?

File: src/controller/BUILD.gn [hunk 2]
```diff
@@ -25,26 +27,31 @@ static_library("controller") {
...
-    "SetUpCodePairer.cpp",
-    "SetUpCodePairer.h",
   ]
 
+  if (chip_python_controller) {
+    sources += [
+      "AbstractDnssdDiscoveryController.cpp",
+      "CHIPCommissionableNodeController.cpp",
+      "CHIPCommissionableNodeController.h",
+      "CHIPDevice.cpp",
```
Comment (bzbarsky-apple) [inline src/controller/BUILD.gn:39]: This definitely looks odd.  Why do we not need CHIPDevice.cpp for _all_ controllers?

File: src/controller/python/chip/server/ServerInit.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    uint16_t securePort   = CHIP_PORT;
+    uint16_t unsecurePort = CHIP_UDC_PORT;
+
+    // Init ZCL Data Model and CHIP App Server
+    chip::Server::GetInstance().Init(nullptr, securePort, unsecurePort);
+
+    // Initialize device attestation config
+    // SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());
+
+    result   = pthread_create(&sPlatformMainThread, nullptr, PlatformMainLoop, nullptr);
```
Comment (bzbarsky-apple) [inline src/controller/python/chip/server/ServerInit.cpp:180]: This is sort of reinventing `StartEventLoopTask`, but OK.

Comment (markus-becker-tridonic-com) [reply src/controller/python/chip/server/ServerInit.cpp:180]: @bzbarsky-apple this is as it is currently in https://github.com/project-chip/connectedhomeip/blob/53dd5833f08e5a6e4f19ed26442b5fb3510cf7ec/src/controller/python/chip/native/StackInit.cpp#L83. If it needs to be changed in this PR, then it should be changed in there as well.

File: src/app/chip_data_model.gni (hunk 1 above)
Comment (markus-becker-tridonic-com) [reply src/app/chip_data_model.gni:157]: By default `chip_python_controller` is true (https://github.com/project-chip/connectedhomeip/pull/11116/files/c1c0be1ccc506bcdbdf148a0ebef853134ea505c#diff-b3871ceca7b7fb48c3fd91178d2c0893f75735547c31eb1f7cac92f0f968f9cbR19). Only for the Python end-device it is set to false.

Possibly I should rename the variable to `chip_controller` to make it clear, that it does not only apply to the Python controller.

File: scripts/build_python_device.sh [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+"
+}
+
+file_name=${0##*/}
+
+while (($#)); do
+    case $1 in
+        --help | -h)
+            help
+            exit 1
```
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply scripts/build_python_device.sh:69]: same as here: https://github.com/project-chip/connectedhomeip/blob/master/scripts/build_python.sh#L69

File: examples/lighting-app/python/lighting.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+dali_loop = None
+dev = None
+
+
+async def dali_on(is_on: bool):
+    global dali_loop
+    global dev
+
+    #print("Waiting for device...")
```
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply examples/lighting-app/python/lighting.py:51]: Those helped my during debugging and to keep up with ToT to ensure that the code still works.

File: src/controller/BUILD.gn (hunk 2 above)
Comment (markus-becker-tridonic-com) [reply src/controller/BUILD.gn:39]: same as above

File: src/app/chip_data_model.gni (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/chip_data_model.gni:162]: Yes, renaming that way would be a lot clearer.

File: src/controller/python/chip/server/ServerInit.cpp (hunk 3 above)
Comment (bzbarsky-apple) [reply src/controller/python/chip/server/ServerInit.cpp:180]: It doesn't need to be changed in this PR, but maybe file a followup issue to look into using `StartEventLoopTask` instead?

Comment (markus-becker-tridonic-com) [reply src/controller/python/chip/server/ServerInit.cpp:180]: https://github.com/project-chip/connectedhomeip/issues/11416

File: src/controller/BUILD.gn (hunk 2 above)
Comment (markus-becker-tridonic-com) [reply src/controller/BUILD.gn]: Done.

File: examples/lighting-app/python/lighting.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+#    limitations under the License.
+#
+
+from chip.server import (
+    GetLibraryHandle,
+    NativeLibraryHandleMethodArguments,
+    PostAttributeChangeCallback,
+)
+
+# from chip.server.types import PostAttributeChangeCallback
```
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply examples/lighting-app/python/lighting.py]: removed

File: examples/lighting-app/python/lighting.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply examples/lighting-app/python/lighting.py]: removed.

File: scripts/build_python_device.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply scripts/build_python_device.sh:69]: keeping as is for the reason above.

**Comment by woody-apple** [comment]
@markus-becker-tridonic-com Please note ZAP is failing for real :)

**Comment by markus-becker-tridonic-com** [comment]
> @markus-becker-tridonic-com Please note ZAP is failing for real :)

rebase seemed to fix zap. not sure whether `Tests / Test Suites - Darwin` and `QEMU / ESP32` are flaky CI tests.

File: examples/lighting-app/python/BUILD.gn [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  include_dirs = [
+    ".",
+    "include",
+    "${chip_root}/zzz_generated/lighting-app/",
+  ]
+}
+
+shared_library("chip-lighting-app") {
+  sources = [
+    #    "${chip_root}/src/controller/python/chip/setup_payload/Generator.cpp",
```
Comment (andy31415) [inline examples/lighting-app/python/BUILD.gn:39]: please do not keep commented-out code, delete it instead.

This looks like an empty shared library - is it needed?

Comment (markus-becker-tridonic-com) [reply examples/lighting-app/python/BUILD.gn]: You're correct. This was a remainder from the earliest tests on where to build the library. Complete file removed.

**Comment by markus-becker-tridonic-com** [comment]
@woody-apple rebased and conflict resolved.

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
Fast tracking given this has had enough time for review.

**Pull request merged**